### PR TITLE
[WEB-1220] Checkout Form Validation

### PR DESCRIPTION
**Pull request by tony-dinh**
Add form validation for checkout flow.

 **JIRA**: https://mobify.atlassian.net/browse/WEB-1220
 **Linked PRs**: N/A

## Changes
- Add basic form validation for checkout flow & tax estimation
- Remove runtype errors in checkout flow
- Remove login button on empty cart if already logged in
- Add CC validation with luhn algorithm & CC expiry validation
- `Calculate` button should appear on cart summary to calculate tax when there is no zip code present.

## Todos
- [ ] +1

## How to test-drive this PR
- Make sure your node_modules is pretty fresh (works with latest develop)
- Run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Run through the checkout flow and notice:
    1. There are no longer runtype validation errors every page
    1. Checkout will not allow you to proceed with empty inputs.
    1. Checkout payment will not let you proceed with a credit card number that does not pass Luhn's checksum algorithm. (a card # that does work: 40303000000... any amount of 0's after will pass)

Existing Issues that were noted on the ticket that are not fixed due to scope:
- Saved credit cart not appearing (logged in) 

Although I don't see anywhere we can save a credit card?.. so maybe this isn't an issue until we have saved credit accounted for.

### CODE REVIEW

File: web/app/containers/cart/partials/cart-summary.jsx [hunk 1]
```diff
@@ -95,7 +95,7 @@ const CartSummary = ({
                         />
                     }
 
-                    {zipCode &&
+                    {(zipCode !== null && zipCode !== undefined) &&
```
Comment (tony-dinh) [inline web/app/containers/cart/partials/cart-summary.jsx]: This is to remove the console log warning that complains about there being whitespace being rendered.

File: web/app/integration-manager/_merlins-connector/checkout/commands.js [hunk 2]
```diff
@@ -105,9 +113,14 @@ const processShippingData = ($response) => (dispatch, getState) => {
...
-    const isInitialized = shippingSelectors.getIsInitialized(getState())
+    const state = getState()
+    const isInitialized = shippingSelectors.getIsInitialized(state)
+
     if (!isInitialized) {
-        dispatch(receiveShippingAddress(parseShippingInitialValues(magentoFieldData)))
+        const initialData = parseShippingInitialValues(magentoFieldData)
+        const currentAddress = shippingSelectors.getShippingAddress(state).toJS()
+
+        dispatch(receiveShippingAddress({...currentAddress, ...initialData}))
```
Comment (jeffkamo) [inline web/app/integration-manager/_merlins-connector/checkout/commands.js]: should `initialData` override `currentAddress` or vice versa?

File: web/app/integration-manager/_sfcc-connector/cart/utils.js [hunk 3]
```diff
@@ -83,20 +83,26 @@ export const fetchCartItemImages = () => (dispatch, getState) => {
...
-                        const product = getProductById(productId)(currentState).toJS()
+                    .then(({image_groups, name, page_title, short_description}) => {
+                        const productHref = getProductHref(productId)
+                        const product = {
+                            id: productId,
+                            title: page_title,
+                            available: true,
+                            href: productHref,
+                            price: '',
+                            ...getProductById(productId)(currentState).toJS()
```
Comment (jeffkamo) [inline web/app/integration-manager/_sfcc-connector/cart/utils.js]: the state product should override the above values?

File: web/app/modals/cart-estimate-shipping/container.jsx [hunk 4]
```diff
@@ -119,4 +169,5 @@ const EstimateShippingReduxForm = ReduxForm.reduxForm({
     form: ESTIMATE_FORM_NAME
 })(CartEstimateShippingModal)
 
+export {CartEstimateShippingModal}
 export default connect(mapStateToProps, mapDispatchToProps)(EstimateShippingReduxForm)
```
Comment (jeffkamo) [inline web/app/modals/cart-estimate-shipping/container.jsx]: Is there a need for both of these exports?

File: web/app/store/form/selectors.js [hunk 5]
```diff
@@ -34,21 +34,7 @@ export const isRegionFreeform = (formName) => createSelector(
 
 export const getFormAddressValues = (formKey) => createSelector(
     getFormValues(formKey),
-    isRegionFreeform(formKey),
```
Comment (jeffkamo) [inline web/app/store/form/selectors.js]: Was this the only place that used `isRegionFreeform `? Can the definition be removed too?

File: web/app/modals/cart-estimate-shipping/container.jsx (hunk 4 above)
Comment (tony-dinh) [inline web/app/modals/cart-estimate-shipping/container.jsx]: This one was a weird one.. lol
One of our tests required the CartEstimateShippingModal

File: web/app/integration-manager/_merlins-connector/checkout/commands.js (hunk 2 above)
Comment (tony-dinh) [inline web/app/integration-manager/_merlins-connector/checkout/commands.js]: Mm misleading... changing that now.
Meant to convey: data from the store needs to be updated/initialized with data from magento

File: web/app/store/form/selectors.js (hunk 5 above)
Comment (tony-dinh) [inline web/app/store/form/selectors.js]: Hm yep seems like nothing else is using it for our demos but it seems to be a part of the interface for IM.

Similar to the other methods defined above it. Nothing using it... but I think it's just meant to be there as part of the interface of IM.

File: web/app/integration-manager/_sfcc-connector/cart/utils.js (hunk 3 above)
Comment (tony-dinh) [inline web/app/integration-manager/_sfcc-connector/cart/utils.js]: Good catch...

**Comment by tony-dinh** [comment]
@jeffkamo Hmm somehow I was able to go through the SFCC test just now... Although Ive struggled getting through it once before where the form wouldn't let me.. but if I later tried.. it worked... this was before this PR though. So idk if that is related to the issue you ran into :O

**Pull request merged**